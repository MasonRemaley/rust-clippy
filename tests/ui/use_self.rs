#![feature(plugin)]
#![plugin(clippy)]
#![warn(use_self)]
#![allow(dead_code)]
#![allow(should_implement_trait)]


fn main() {}

mod use_self {
    struct Foo {}

    impl Foo {
        fn new() -> Foo {
            Foo {}
        }
        fn test() -> Foo {
            Foo::new()
        }
    }

    impl Default for Foo {
        fn default() -> Foo {
            Foo::new()
        }
    }
}

mod better {
    struct Foo {}

    impl Foo {
        fn new() -> Self {
            Self {}
        }
        fn test() -> Self {
            Self::new()
        }
    }

    impl Default for Foo {
        fn default() -> Self {
            Self::new()
        }
    }
}

mod lifetimes {
    struct Foo<'a>{foo_str: &'a str}

    impl<'a> Foo<'a> {
        // Cannot use `Self` as return type, because the function is actually `fn foo<'b>(s: &'b str) -> Foo<'b>`
        fn foo(s: &str) -> Foo {
            Foo { foo_str: s }
        }

        // cannot replace with `Self`, because that's `Foo<'a>`
        fn bar() -> Foo<'static> {
            Foo { foo_str: "foo"}
        }

        // `Self` is applicable here
        fn clone(&self) -> Foo<'a> {
            Foo {foo_str: self.foo_str}
        }
    }
}

mod generics {
    struct Foo<T> {
        value: T,
    }

    impl<T> Foo<T> {
        // `Self` is applicable here
        fn foo(value: T) -> Foo<T> {
            Foo { value }
        }

        // `Cannot` use `Self` as a return type as the generic types are different
        fn bar(value: i32) -> Foo<i32> {
            Foo { value }
        }
    }
}
