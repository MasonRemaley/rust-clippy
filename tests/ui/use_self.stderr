error: unnecessary structure name repetition
  --> $DIR/use_self.rs:14:21
   |
14 |         fn new() -> Foo {
   |                     ^^^ help: use the applicable keyword: `Self`
   |
   = note: `-D use-self` implied by `-D warnings`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:15:13
   |
15 |             Foo {}
   |             ^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:17:22
   |
17 |         fn test() -> Foo {
   |                      ^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:18:13
   |
18 |             Foo::new()
   |             ^^^^^^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:23:25
   |
23 |         fn default() -> Foo {
   |                         ^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:24:13
   |
24 |             Foo::new()
   |             ^^^^^^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:63:28
   |
63 |         fn clone(&self) -> Foo<'a> {
   |                            ^^^^^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:64:13
   |
64 |             Foo {foo_str: self.foo_str}
   |             ^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:76:29
   |
76 |         fn foo(value: T) -> Foo<T> {
   |                             ^^^^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:77:13
   |
77 |             Foo { value }
   |             ^^^ help: use the applicable keyword: `Self`

error: aborting due to 10 previous errors

